File: backend/src/Application/Examinations/ExaminationDetails.cs
using FluentValidation;

namespace Application.Examinations;

/// <summary>
/// The clinical payload of an examination — what the vet types in. Shared by completing an
/// appointment, recording a walk-in, and editing afterwards, so its rules exist once.
/// </summary>
public sealed record ExaminationDetails(
    string PerformedByFirstName,
    string PerformedByLastName,
    string? Anamnesis,
    string? Diagnosis,
    string? Therapy,
    decimal? Cost);

internal sealed class ExaminationDetailsValidator : AbstractValidator<ExaminationDetails>
{
    public ExaminationDetailsValidator()
    {
        RuleFor(d => d.PerformedByFirstName).NotEmpty().MaximumLength(100);
        RuleFor(d => d.PerformedByLastName).NotEmpty().MaximumLength(100);
        RuleFor(d => d.Anamnesis).MaximumLength(4000);
        RuleFor(d => d.Diagnosis).MaximumLength(4000);
        RuleFor(d => d.Therapy).MaximumLength(4000);
        RuleFor(d => d.Cost).GreaterThanOrEqualTo(0).When(d => d.Cost.HasValue);
    }
}

File: frontend/src/features/examinations/lib/examinationDetails.ts
import type { ExaminationDetails, ExaminationFormValues } from '../types'

function trimmed(value: string): string | undefined {
  const text = value.trim()
  return text ? text : undefined
}

export function parseCost(value: string): number | undefined {
  const text = value.trim().replace(',', '.')
  if (!text) return undefined

  const cost = Number(text)
  return Number.isFinite(cost) ? cost : undefined
}

export function toExaminationDetails(values: ExaminationFormValues): ExaminationDetails {
  return {
    performedByFirstName: values.performedByFirstName.trim(),
    performedByLastName: values.performedByLastName.trim(),
    anamnesis: trimmed(values.anamnesis),
    diagnosis: trimmed(values.diagnosis),
    therapy: trimmed(values.therapy),
    cost: parseCost(values.cost),
  }
}

export function emptyExaminationValues(
  performedByFirstName = '',
  performedByLastName = '',
): ExaminationFormValues {
  return {
    performedByFirstName,
    performedByLastName,
    anamnesis: '',
    diagnosis: '',
    therapy: '',
    cost: '',
  }
}
